### PR TITLE
Validate that handlers don't use message session

### LINKED ISSUE #3749: V6 - Prevent users from injecting IEndpointInstance and using it inside of Message handlers to send / publish messages
Relates to: 
https://github.com/Particular/V6Launch/issues/27
https://github.com/Particular/PlatformDevelopment/issues/819

### Problem

Often times customers have old / common libraries where a IBus was injected into the library. And the IBus was used to send messages / publish events from a self host or a NServiceBus.Host either in bootstrap scenarios or in message handlers. 

However, `IEndpointInstance/MessageSession` is not safe to use from message handlers. 
Some example of issues with sending messages using IEndpointInstance/IMessageSession within a message handler are:
- those messages will not participate in the same transaction scope as that of the message handler. This could result in messages dispatched or events published via the `IEndpointInstance` interface even if the message handler resulted in an exception and the operation was rolled back.
- those messages will not be part of the [batching operation](/nservicebus/messaging/batched-dispatch.md) and therefore will not have the benefits that Version 6 provides.
- those messages will not contain any important message header information that is available via the `IHandlerMessageContext` interface parameter, e.g., CorrelationId.  

How can we stop new and existing users from writing "invalid" code from either compiling or throwing a run time warning / exception?

Because without it and just relying on guidance, users can still run into issues with code that compile and seems like it runs fine on the surface while leading to undesirable run time downsides, especially when a message handler transaction fails.

### Proposed solution

**Type checking on handlers at startup**
To prevent users from wrongly using IEndpointInstance and IMessageSession on their handlers we should check on startup and throw an exception if any handler has a ctor/property dependency in said types. The exception should tell them to use the handler context instead.

### PR DESCRIPTION
**Pull request by andreasohlund**
Closes #3749

### CODE REVIEW

**Comment by andreasohlund** [comment]
@Particular/nservicebus-maintainers please review

@indualagarsamy what do you think about the error message? can we improve?

File: src/NServiceBus.Core.Tests/Handlers/MessageHandlerRegistryTests.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            }
+
+            public Task Handle(MyMessage message, IMessageHandlerContext context)
+            {
+                throw new NotImplementedException();
+            }
+
+            IMessageSession MessageSession;
+        }
+        class HandlerWithIEndpointInstanceCtorDep : IHandleMessages<MyMessage>
```
Comment (bording) [inline src/NServiceBus.Core.Tests/Handlers/MessageHandlerRegistryTests.cs]: minor: missing a space here.

File: src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs [hunk 2]
```diff
@@ -137,18 +140,32 @@ static void CacheMethod(Type handler, Type messageType, Type interfaceGenericTyp
...
+                        typeof(IHandleMessages<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t) ||
+                        typeof(IHandleTimeouts<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t)
+                    select potentialMessageType)
                 .Distinct()
                 .ToList();
         }
+        void ValidateHandlerType(Type handlerType)
+        {
+            var propertyTypes = handlerType.GetProperties().Select(p => p.PropertyType).ToList();
+            var ctorArguments = handlerType.GetConstructors()
```
Comment (bording) [inline src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs]: You said that you "check for both private and public properties/ctors" in the PR message, but these overloads are only going to return public properties and constructors.

**Comment by bording** [comment]
If we're testing the `MessageHandlerRegistry` directly, is there value in also having the acceptance test?

Also, would it be worth testing all the permutations for Sagas instead of just the one? I know they go through the same code path, so I guess that isn't worth doing?

File: src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs [hunk 3]
```diff
@@ -137,18 +140,32 @@ static void CacheMethod(Type handler, Type messageType, Type interfaceGenericTyp
...
+             .SelectMany(ctor => ctor.GetParameters().Select(p => p.ParameterType))
+             .ToList();
+
+            var dependencies = propertyTypes.Concat(ctorArguments).ToList();
+
+            if (dependencies.Any(t => typeof(IMessageSession).IsAssignableFrom(t)))
+            {
+                throw new Exception($"IMessageSession/IEndpointInstance shouldn't be injected into message handlers and Sagas. Use the context parameter on the `{handlerType.Name}.Handle` method to send messages.");
```
Comment (indualagarsamy) [inline src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs]: @andreasohlund - Looks good. Minor tweaks.  How about:

Interfaces IMessageSession or IEndpointInstance should not be registered in the container to enable sending or publishing messages from within sagas or message handlers.  Instead, use the context parameter on the `{handlerType.Name}.Handle` method to send or publish messages.

File: src/NServiceBus.Core.Tests/Handlers/MessageHandlerRegistryTests.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        public void ShouldThrowIfUserTriesToBypassTheHandlerContext(Type handlerType)
+        {
+            var registry = new MessageHandlerRegistry(new Conventions());
+
+            Assert.Throws<Exception>(() => registry.RegisterHandler(handlerType));
+        }
+
+        class HandlerWithIMessageSessionProperty : IHandleMessages<MyMessage>
+        {
+            public IMessageSession MessageSession { get; set; }
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Handlers/MessageHandlerRegistryTests.cs]: Do we need a test for a property of type `IMessageSession` on a base class?

File: src/NServiceBus.Core.Tests/Handlers/MessageHandlerRegistryTests.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+            public Task Handle(MyMessage message, IMessageHandlerContext context)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        class HandlerWithIEndpointInstanceProperty : IHandleMessages<MyMessage>
+        {
+            public IEndpointInstance EndpointInstance { get; set; }
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Handlers/MessageHandlerRegistryTests.cs]: Do we need a test for a property of type `IEndpointInstance ` on a base class?

File: src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs (hunk 2 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs]: If a user uses a container to inject it using private property injection usually requires fiddling with the defaults of the container. I think we are pretty safe if we just check for public properties

Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs]: But I think we should flatten the hierarchy for properties

https://msdn.microsoft.com/en-us/library/kyaxdd3x(v=vs.110).aspx

Comment (andreasohlund) [inline src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs]: >You said that you "check for both private and public properties/ctors" in the PR message, but these overloads are only going to return public properties and constructors.

Duh, exactly what I wanted, updated the description of the PR

**Comment by andreasohlund** [comment]
>If we're testing the MessageHandlerRegistry directly, is there value in also having the acceptance test?

Me and @timbussmann talked about this and we felt that having a single acceptance test that verifies that the check where indeed happening at start up (and not later ) was appropriate. And the cover the edge cases with unit tests. Does this make sense?

File: src/NServiceBus.Core.Tests/Handlers/MessageHandlerRegistryTests.cs (hunk 4 above)
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Handlers/MessageHandlerRegistryTests.cs]: :+1: added

File: src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs (hunk 3 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs]: Updated

Comment (andreasohlund) [inline src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs]: Actually, registering them in the container is fine, injecting them into handlers/sagas is not?

Perhaps reword to 

>Interfaces IMessageSession or IEndpointInstance should not be resolved from the container to enable sending or publishing messages from within sagas or message handlers. Instead, use the context parameter on the {handlerType.Name}.Handle method to send or publish messages.

Thoughts?

File: src/NServiceBus.Core.Tests/Handlers/MessageHandlerRegistryTests.cs (hunk 5 above)
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Handlers/MessageHandlerRegistryTests.cs]: I added one for IMessageSession , that should cover it?

File: src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs (hunk 2 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs]: Flatten doesn't seems to be needed =>https://github.com/Particular/NServiceBus/pull/3862/commits/3f93010d83d6119b70cc541e5d8f840c5eaa0440

that test is green without it?

**Comment by andreasohlund** [comment]
All comments should now be addressed

**Comment by danielmarbach** [comment]
All good from my side

**Comment by danielmarbach** [comment]
@bording your comments seem addressed. Merging
